### DISCUSSION TITLE
ID3v2 metadata tags

### DISCUSSION THREAD

**Discussion by iSerganov** [savonet/liquidsoap#4080] [Q&A] [answered]
I am trying to insert custom metadata on metadata change event according to the tutorial example I found somewhere:

```
def set_meta(request, response) =
  # Filter out unusual metadata
  meta = metadata.export(request.query)

  # Grab the returned message
  ret =
    if meta != [] then
      radio.insert_metadata(meta)
      "OK!"
    else
      "No metadata to add!"
  end

  response.html("<html><body><b>#{ret}</b></body></html>")
end
```
It works fine only for a few metadata tags like `artist`, `title` and `year`.
The others I guess are filtered here:
```
# Filter out unusual metadata
  meta = metadata.export(request.query)
```
However ID3v2.4 metadata tags set contains much more possible tags, so I am trying to understand why they are not allowed?

Could you please help to find a way to provide more ID3v2 tags and values in the output Liquidsoap stream?

**Comment by gAlleb** [comment]
Hi, @iSerganov . Yeah, `metadata.export` filters metadata as per `settings.encoder.metadata.export` setting. And I assume it has some defaults -> https://www.liquidsoap.info/doc-dev/settings.html

`settings.encoder.metadata.export := ["artist", "title", "album", "genre", "date", "tracknumber", "comment", "track", "year", "dj", "next", "apic", "pic", "metadata_url", "metadata_block_picture", "coverart"]`

```ruby
def metadata.export(m) =
  exported_keys = settings.encoder.metadata.export()
  list.assoc.filter((fun (k, (_:string)) -> list.mem(k, exported_keys)), m)
end
```

<hr>

You can insert whatever metadata you need. Just don't filter it with `metadata.export` function or expand `settings.encoder.metadata.export` - `settings.encoder.metadata.export := ["artist", "title", "WTF", ... ... ...]`.

```ruby
# here we don't use `metadata.export` — thus all metadata is allowed to pass through.

radio = insert_metadata(radio)
def set_metadata(request)
log.important("Inserting metadata: #{request.query}")
radio.insert_metadata(request.query)
http.response(data="Done.")
end
harbor.http.register.simple("/set_metadata", set_metadata, port=8007, method="GET")

meta = ref([])
radio.on_metadata(fun (m) -> meta := m)
def get_all_meta(_, response) =
  response.json(meta())
end
harbor.http.register(port=8007, method="GET", "/metadata", get_all_meta)

```

Now if we do something like `http://localhost:8007/set_metadata?artist=1&title=2&WTF=3` - metadata will change and if you go to `http://localhost:8007/metadata` you'll see something like:

```ruby
[["WTF","3"],["artist","1"],["title","2"]]
```

So basically you have to change `meta = metadata.export(request.query)` to `meta = request.query` to remove filter or apply new settings via `settings.encoder.metadata.export`. Both options are valid.

**Reply by iSerganov** [reply]
Thank you - that works fine!

Just wondering how can I propagate this custom metadata further to Icecast?
At the moment I can see only few tags when I `ffprobe` my output Icecast stream (where Liquidsoap is source client):
```
ffprobe -i http://my.stream/ice/radio.mp3

Input #0, mp3, from 'http://my.stream/ice/radio.mp3':
  Metadata:
    icy-name        : Nice Broadcast
    icy-pub         : 1
    StreamTitle     : Artist - Title
  Duration: N/A, start: 0.000000, bitrate: 192 kb/s
  Stream #0:0: Audio: mp3 (mp3float), 44100 Hz, stereo, fltp, 192 kb/s
```
Is that something that can be configured on Liquidsoap side or is it purely Icecast business?

**Reply by gAlleb** [reply]
This will depend on the format. If we are talking about `Icecast` - It is how `mp3 or aac` icecast stream works. More legacy support. Less everything) It's typically only title, or artist & title. Good old `StreamTitle`. LS uses `ICY metadata` mechanism to update metadata in icecast’s streams because `mp3 or aac` icecast streams do not support in-stream metadata.

Lots of web radios try to put everything into this `StreamTitle`.  Like:

```ruby
def apply_meta(m) =
title = m["title"]
artist = m["artist"]
album = m["album"]
if artist != "" and title != "" then
[...list.assoc.remove("artist", m), ("title", "#{artist} - #{title} - #{album}")]
else
m
end
end
```

Looks weird.

E.g. There is `ogg` stream. It is the container format and holds more metadata. You can try it and ffprobe it to see the difference. Create an icecast output `output.icecast(%ffmpeg(format="ogg", %audio(codec="libopus", samplerate=48000)), host = "", port = 8000, password = "", mount = "ogg",  name="", radio)`. Also `ogg` is a little different as a stream in a way that It's chaining the files. And also It is entirely up to the software to display this metadata in a way it can (e.g. VLC can). 

Best approach will be probably to use a dedicated API and a Player. To grab metadata directly from LS and insert it into Player.

**Reply by iSerganov** [reply]
Makes sense. So `StreamTitle` is built from `"#{artist} - #{title} - #{album}"`
What about `StreamURL`? How can I pass album art link there?

Also it looks like I am doing something wrong when trying to insert custom metadata per each track.
I have slightly modified tour above example:

```
radio = insert_metadata(radio)
def set_metadata(request)
log.important("Inserting metadata: #{request.query}")
radio.insert_metadata(request.query)
http.response(data="Done.")
end
harbor.http.register.simple("/set_metadata", set_metadata, port=8007, method="GET")

meta = ref([])
radio.on_metadata(fun (m) -> meta := m)
def get_all_meta(_, response) =
  response.json(meta())
end

# modify metadata for each track
def apply_metadata(m) =
  title = m["title"]
  artist = m["artist"]
  print("#{artist} - #{title}")
  # Return
  [
    ("isrc","some dynamic code"),
    ("EncodedBy","Nice Encoder")
  ]
end
radio = metadata.map(apply_metadata, radio)
harbor.http.register(port=8007, method="GET", "/metadata", get_all_meta)
```

When I call `/metadata` I do not see new metadata fields. Can you please help?

**Comment by gAlleb** [comment]
> StreamURL? How can I pass album art link there?

Not sure about it. `StreamURL` is an `icy-url` and can be set in the output: `output.icecast(%ffmpeg(format="mp3", %audio(codec="libmp3lame")), host = "", port = 8000, password = "", mount = "mp3", url = "https://URL" , radio)`
icy-name, icy-url are sent at the beginning of the connection and supposedly don't get updated. 

> When I call /metadata I do not see new metadata fields. Can you please help?

To modify metadata you have to go to `http://localhost:8007/set_metadata?artist=1&title=2` and modify it this way.

I presume you want to add this `("isrc","some dynamic code"), ("EncodedBy","Nice Encoder")` to your meta?

Then it's much easier to annotate your playlists:

```ruby
radio = playlist(prefix='annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":', "/music/music.m3u8")

meta = ref([])
radio.on_metadata(fun (m) -> meta := m)
def get_all_meta(_, response) =
  response.json(meta())
end

harbor.http.register(port=8007, method="GET", "/metadata", get_all_meta)

```

Then if needed you can add `set_metadata` function. But if you just want to add `new metadata` - you can annotate inside playlist operator.

**Reply by smoralis** [reply]
> > StreamURL? How can I pass album art link there?
> 
> Not sure about it. `StreamURL` is an `icy-url` and can be set in the output: `output.icecast(%ffmpeg(format="mp3", %audio(codec="libmp3lame")), host = "", port = 8000, password = "", mount = "mp3", url = "https://URL" , radio)` icy-name, icy-url are sent at the beginning of the connection and supposedly don't get updated.
> 
> > When I call /metadata I do not see new metadata fields. Can you please help?
> 
> To modify metadata you have to go to `http://localhost:8007/set_metadata?artist=1&title=2` and modify it this way.
> 
> I presume you want to add this `("isrc","some dynamic code"), ("EncodedBy","Nice Encoder")` to your meta? (Now your `apply_metadata(m)` function does nothing.)
> 
> Then it's much easier to annotate your playlists:
> 
> ```ruby
> radio = playlist(prefix='annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":', "/music/music.m3u8")
> 
> meta = ref([])
> radio.on_metadata(fun (m) -> meta := m)
> def get_all_meta(_, response) =
>   response.json(meta())
> end
> 
> harbor.http.register(port=8007, method="GET", "/metadata", get_all_meta)
> ```
> 
> Then if needed you can add `set_metadata` function. But if you just want to add `new metadata` - you can annotate inside playlist operator.

icy-url is as said the url of the website and is set as described and doesn't get updated. 
StreamURL is different, like StreamTitle and can get updated. 
To add the url of the artwork add to for example 
http://localhost:8007/set_metadata?artist=1&title=2&url=url_of_image
Icecast will convert the url_of_image to StreamURL (applies only to legacy mp3 or aac streams)

**Reply by gAlleb** [reply]
This will just set and update "url" tag that won't be sent to icecast. No?
I'm unable to reproduce the behavior you've described.

**Reply by smoralis** [reply]
> This will just set and update "url" tag that won't be sent to icecast. No? I'm unable to reproduce the behavior you've described.

Does your settings.encoder.metadata.export include the value "url"?

**Reply by gAlleb** [reply]
They do. But can you reproduce it?

`http://localhost:8007/set_metadata?artist=1&title=2&url=url_of_image` will just set and update "url" tag. 
StreamURL - is an icy-url which you can see in ffprobe. "Url" tag metadata possibly won't be sent to icecast.

**Reply by smoralis** [reply]
> They do. But can you reproduce it?
> 
> `http://localhost:8007/set_metadata?artist=1&title=2&url=url_of_image` will just set and update "url" tag. StreamURL - is an icy-url which you can see in ffprobe. "Url" tag metadata possibly won't be sent to icecast.

Yes
Here is my script
```
log.stdout.set(true)
log.file.set(true)

settings.encoder.metadata.export := ["artist", "title", "album", "genre", "date", "tracknumber", "comment", "track", "year", "dj", "next", "apic", "pic", "metadata_url", "metadata_block_picture", "coverart", "cover_url" , "url"]

settings.harbor.bind_addrs := ["0.0.0.0"]

s = mksafe(buffer(input.portaudio(device_id=3)))

s = insert_metadata(s)

def set_metadata(request)
log.important("Inserting metadata: #{request.query}")
s.insert_metadata(request.query)
http.response(data="Done.")
end

harbor.http.register.simple(port = 8090, "/set_metadata", set_metadata)


#Icecast

genre = "Experimental Radio"
description = "Experimental Streaming Testing"
host = "zeremy.serv00.net"
port = 8201
url = "https://zeremy.serv00.net"
password = 

#aac
 output.icecast(%ffmpeg(
  format="adts",
  %audio(samplerate = 44100, codec = "aac" , b = "320k")),
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [AAC 320]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "aac",
  send_icy_metadata=true,
  icy_metadata=["song", "title", "artist", "url"],
  s)

#mp3  
output.icecast(%ffmpeg(
  format="mp3",
  %audio(samplerate = 44100, codec = "libmp3lame", b = "320k")),  
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [MP3 320]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "mp3",
  send_icy_metadata=true,
  icy_metadata=["song", "title", "artist", "url"],
  s)
  
#opus
output.icecast(%ffmpeg(
  format="ogg",
  %audio(samplerate = 48000, codec = "libopus", b = "128k")),
  host = host, 
  port = port,
  fallible=false,
  name= "LSFB Experimental Radio [OPUS CHAINED 128]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "opus",
  send_icy_metadata=false,
  s)
  
  #opus RSAS
output.icecast(%ffmpeg(
  format="ogg",
  %audio(samplerate = 48000, codec = "libopus", b = "128k")),
  host = host, 
  port = 8202,
  fallible=false,
  name= "LSFB Experimental Radio [OPUS ICY 128]",
  description = description,
  genre = genre,
  url = url,
  password = password,
  mount= "opus",
  send_icy_metadata=false,
  source.drop.metadata(s))
 ```

You can see the metadata sent
https://eshaz.github.io/icecast-metadata-js/demo.html

Add endpoint
https://zeremy.serv00.net:8201/mp3

**Reply by iSerganov** [reply]
Thanks for your answers gentlemen! They are very helpful, only 1 thing left for me to understand.

@gAlleb , playlist annotation would have worked for me if the metadata tag values were static, but they are different from song to song.
E.g., an input track may contain `isrc`: `<unique_song_code>` metadata tag, however for many other tracks I would like to insert it on the fly and I need to update metadata automatically when the next song starts playing (without making manual API call).
Is it something I can do with Liquidsoap?

**Reply by gAlleb** [reply]
Thank you, @smoralis! I've been able to reproduce it)

```
[mp3 @ 0x7f8990f04b40] Skipping 264 bytes of junk at 0.
Input #0, mp3, from 'http://localhost:8000/mac.mp3':
  Metadata:
    icy-br          : 0
    icy-description : For meditation and relaxation
    icy-genre       : It's music for your ears
    icy-name        : omFM
    icy-pub         : 1
    icy-url         : pdf
    StreamTitle     : 1 - 2
    StreamUrl       : test
```
Ha! But here a little note should be made: that it has to be `Icecast 2.5` :) With 2.4.4 it doesn't work.
<img width="400" alt="image" src="https://github.com/user-attachments/assets/a0757e02-6364-4bfe-b52b-24d8a66c2c02">

**Reply by gAlleb** [reply]
@iSerganov, yeah, it's possible.

> E.g., an input track may contain isrc: <unique_song_code> metadata tag. 

You mean that the track already contains this meta? Then you don't have to update anything. If it has metadata - then it's available. You can include it into `settings.encoder.metadata.export` to be sure.

If your track don't contain this meta — you can either annotate every track in playlist:

```
#Playist.m3u:

annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
```

Or else you need a more complex function that will fetch needed metadata and inject it on track change on-the-fly. 
But it's easier maybe to write a small script to annotate or inject tags beforehand.

And again: if tracks already contains this tag - you don't have to update anything. Tags will be read by LS.

**Reply by smoralis** [reply]
> Thank you, @smoralis! I've been able to reproduce it)
> 
> ```
> [mp3 @ 0x7f8990f04b40] Skipping 264 bytes of junk at 0.
> Input #0, mp3, from 'http://localhost:8000/mac.mp3':
>   Metadata:
>     icy-br          : 0
>     icy-description : For meditation and relaxation
>     icy-genre       : It's music for your ears
>     icy-name        : omFM
>     icy-pub         : 1
>     icy-url         : pdf
>     StreamTitle     : 1 - 2
>     StreamUrl       : test
> ```
> 
> Ha! But here a little note should be made: that it has to be `Icecast 2.5` :) With 2.4.4 it doesn't work. <img alt="image" width="400" src="https://private-user-images.githubusercontent.com/144447266/355995358-a0757e02-6364-4bfe-b52b-24d8a66c2c02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNjczMTUsIm5iZiI6MTcyMzA2NzAxNSwicGF0aCI6Ii8xNDQ0NDcyNjYvMzU1OTk1MzU4LWEwNzU3ZTAyLTYzNjQtNGJmZS1iNTJiLTI0ZDhhNjZjMmMwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QyMTQzMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMTNmMDUwMDc1YzQyYWEyNGQ1YTdlYTk2NTA4MTNkMWU2MDIxMmU1NjU5YmIwODRhZTc3MDg0NWM1ZjZjNDFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QcnEgGk7PP7GzZZpuGrNCqqEoWA2H0D71hxjW8kb3Ro">

Ok, great!
I can't test 2.4.4 at the moment but the code is in the master.
https://github.com/xiph/Icecast-Server/blob/57093def7baacf2aaaff3a17915cd7fc0b8cfec3/src/format_mp3.c#L245

I'm sure that with IcecastKH it is also possible.

**Reply by iSerganov** [reply]
> @iSerganov, yeah, it's possible.
> 
> > E.g., an input track may contain isrc: <unique_song_code> metadata tag.
> 
> You mean that the track already contains this meta? Then you don't have to update anything. If it has metadata - then it's available. You can include it into `settings.encoder.metadata.export` to be sure.
> 
> If your track don't contain this meta — you can either annotate every track in playlist:
> 
> ```
> #Playist.m3u:
> 
> annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
> annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
> annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
> ```
> 
> Or else you need a more complex function that will fetch needed metadata and inject it on track change on-the-fly.

Can you provide me with a small example of the function to change metadata on the fly please?

**Reply by iSerganov** [reply]
Oh, and another question (sorry, today it seems too many questions :))
@gAlleb, you have mentioned:

```
#Playist.m3u:

annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":/path/to/track
```
Can I do the same with `.pls` playlist? E.g.:
```
...
annotate:isrc="track5 ISRC Code",EncodedBy="Nice Encoder"
File5=https://path.to.track/track5.mp3
Title5=Nice name
Length5=277
...
```

**Reply by smoralis** [reply]
> > Thank you, @smoralis! I've been able to reproduce it)
> > ```
> > [mp3 @ 0x7f8990f04b40] Skipping 264 bytes of junk at 0.
> > Input #0, mp3, from 'http://localhost:8000/mac.mp3':
> >   Metadata:
> >     icy-br          : 0
> >     icy-description : For meditation and relaxation
> >     icy-genre       : It's music for your ears
> >     icy-name        : omFM
> >     icy-pub         : 1
> >     icy-url         : pdf
> >     StreamTitle     : 1 - 2
> >     StreamUrl       : test
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Ha! But here a little note should be made: that it has to be `Icecast 2.5` :) With 2.4.4 it doesn't work. <img alt="image" width="400" src="https://private-user-images.githubusercontent.com/144447266/355995358-a0757e02-6364-4bfe-b52b-24d8a66c2c02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNjczMTUsIm5iZiI6MTcyMzA2NzAxNSwicGF0aCI6Ii8xNDQ0NDcyNjYvMzU1OTk1MzU4LWEwNzU3ZTAyLTYzNjQtNGJmZS1iNTJiLTI0ZDhhNjZjMmMwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QyMTQzMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMTNmMDUwMDc1YzQyYWEyNGQ1YTdlYTk2NTA4MTNkMWU2MDIxMmU1NjU5YmIwODRhZTc3MDg0NWM1ZjZjNDFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QcnEgGk7PP7GzZZpuGrNCqqEoWA2H0D71hxjW8kb3Ro">
> 
> Ok, great! I can't test 2.4.4 at the moment but the code is in the master. https://github.com/xiph/Icecast-Server/blob/57093def7baacf2aaaff3a17915cd7fc0b8cfec3/src/format_mp3.c#L245
> 
> I'm sure that with IcecastKH it is also possible.

@gAlleb 
I can confirm 2.4.4 doesn't work.
The commit fixing admin.c is https://gitlab.xiph.org/xiph/icecast-server/-/commit/72fbe999e29e7ebf5fbbf3eabbc33e1b8f2c2d2b

**Reply by gAlleb** [reply]
Thanks, @smoralis!

**Comment by gAlleb** [comment]
> Can I do the same with .pls playlist? E.g.:

`annotate` should be before the path. If it's `https://path.to.track/track5.mp3` - then something like `annotate:isrc="some dynamic code",EncodedBy="Nice Encoder":https://path.to.track/track5.mp3` will work with a basic playlist. Not sure about the above format. You can test.

> Can you provide me with a small example of the function to change metadata on the fly please?

It's still not clear to me what the goal is exactly. It's always better to do this beforehand and edit your songs with necessary tags.

1. On track change we either insert new metadata - the result will be: only metadata that we have passed:

```
radio = insert_metadata(radio)

def set_metadata(_)

radio.insert_metadata([("anytag", "q"),("anytag", "q")])

end

radio.on_track(set_metadata)
```

2. Or we can append some new metadata to the existing one.

```
def append_meta(_) =
  [("isrc","1"),("EncodedBy","2")]
end
radio = metadata.map(append_meta, radio)
```

3. Or combine. We use `ref`, then `on_track` event we populate them with strings that we need and append this metadata

```
isrc = ref("")
EncodedBy = ref("")

def set_metadata(_)
# get data from somewhere and populate refs
isrc :=  "test"
EncodedBy :=  "test2"

end

radio.on_track(set_metadata)

def append_meta(_) =
  [("isrc", isrc()),("EncodedBy", EncodedBy())]
end

radio = metadata.map(append_meta, radio)
```

**Reply by gAlleb** [reply, marked as answer]
4. If it is an option - you can setup a dedicated API for this, which would load the same playlist and assign new metadata to every track respectfully or keep this data as an array alongside `artist - title` values or DB of some sort. Then on every new track you can send  current `artist - title` value to the API and in response get `new meta` (taken from an array) which you will then append:

```ruby

# Create `refs`
new_isrc = ref("")
new_EncodedBy = ref("")

api_url = "http://api_url"

# Main function

def set_metadata(_)

  # API function with PUT request
  def api_put(path, data=null())

   full_url = "#{api_url}/#{path}"

     log("API #{path} - Sending PUT request to '#{full_url}' with Artist and Title: #{data}")

      try
        response = http.put(full_url,
            headers=[
                ("Content-Type", "application/json")
            ],
       
            data=json.stringify(data, compact=true)
        )
        log("API #{path} - Response (#{response.status_code}): #{response}")

	response

     catch error do
  	null()
     end
   end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : ""
  end

 # We take meta from the source 
  m = radio.last_metadata() ?? [] 

  artist = null_list("artist", m)
  
  title = null_list("title", m)

  # We send Artist - Title 
  get_more_meta = api_put("meta", { artist = "#{artist}", title = "#{title}" })

  if get_more_meta != null() then

  try
   
   # Parse response
    let json.parse ({isrc, EncodedBy} : {isrc: string, EncodedBy: string}) = "#{get_more_meta}"

  # Assign new values
    new_isrc := "#{isrc}"
    new_EncodedBy := "#{EncodedBy}"
   
  catch err do
    log("#{err}")
  end
  end
end

radio.on_track(set_metadata)

# Append new meta to the stream with values we got in the function above.

def append_meta(_) =

[("isrc", new_isrc()),("EncodedBy", new_EncodedBy())]

end

radio = metadata.map(append_meta, radio)

```

_overcomplexity: achieved_

**Reply by iSerganov** [reply]
This seems more or less what I was looking for. 

So the easiest way is to pre-process all input tracks and set all metadata I need there.
However it is always nice to have an 'overcomplicated' plan B which can save me in case if plan A is impossible.

Thank you very much!
